cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java

```diff
@@ -0,0 +1,250 @@
+ * Git hosting services and provides version control operations.
+ */
+public class RepositoryManager implements AutoCloseable {
+  private static final Logger LOG = LoggerFactory.getLogger(RepositoryManager.class);
+  private static final String VALIDATION_REPOSITORY_DIRECTORY_NAME = "validation";
+  private static final String LOCAL_REPOSITORY_DIRECTORY_NAME = "local";
+  // Default value for git timeout.
+  private static final int GIT_COMMAND_TIMEOUT_SECONDS_DEFAULT = 30;
+  // Default value for git repo
+  private static final String GIT_REPOSITORIES_CLONE_DIRECTORY_PATH_DEFAULT = "/tmp/source-control/";
+  private final AuthStrategy authStrategy;
+  private final File localRepoDirectory;
+  private final File validationRepoDirectory;
+  private final int gitCommandTimeoutSeconds;
+  private final SecureStore secureStore;
+  private final Path repoBasePath;
+  private final RepositoryConfig repositoryConfig;
+  private final NamespaceId namespaceId;
+  private Git git;
+
+  public RepositoryManager(SecureStore secureStore,
+                           RepositoryConfig repositoryConfig,
+                           NamespaceId namespaceId,
+                           CConfiguration cConf) {
+    String gitCloneDirectory = cConf.get(Constants.SourceControlManagement.GIT_REPOSITORIES_CLONE_DIRECTORY_PATH,
+                                         GIT_REPOSITORIES_CLONE_DIRECTORY_PATH_DEFAULT);
+    this.secureStore = secureStore;
+    this.repositoryConfig = repositoryConfig;
+    this.namespaceId = namespaceId;
+    this.gitCommandTimeoutSeconds =
+      cConf.getInt(Constants.SourceControlManagement.GIT_COMMAND_TIMEOUT_SECONDS, GIT_COMMAND_TIMEOUT_SECONDS_DEFAULT);
+    this.repoBasePath = new File(gitCloneDirectory).toPath().resolve("namespace").resolve(namespaceId.getNamespace());
+    this.localRepoDirectory = repoBasePath.resolve(LOCAL_REPOSITORY_DIRECTORY_NAME).toFile();
+    this.validationRepoDirectory = repoBasePath.resolve(VALIDATION_REPOSITORY_DIRECTORY_NAME).toFile();
+    try {
+      this.authStrategy = getAuthStrategy();
+    } catch (AuthStrategyNotFoundException e) {
+      //  This exception isn't expected as the API handlers only store auth config after validations,
+      throw new RuntimeException(e);
+    }
+  }
+
+  private <C extends TransportCommand> C getAuthenticatedCommandWithTimeout(C command) throws AuthenticationException {
+    command.setCredentialsProvider(authStrategy.getCredentialProvider(secureStore, repositoryConfig, namespaceId));
+    command.setTimeout(gitCommandTimeoutSeconds);
+    return command;
+  }
+
+  /**
+   * Initializes the Git repository by cloning remote.
+   *
+   * @throws GitAPIException         when a Git operation fails.
+   * @throws AuthenticationException when correct authentication credentials are not available.
+   * @throws IOException             when file or network I/O fails.
+   */
+  public void initialize() throws GitAPIException, AuthenticationException, IOException {
```

Comment:
For class that needs initialize and close, implements the guava `Service` instead (probably extends from `AbstractIdleService` would be appropriate for this class).